src/main/java/org/spongepowered/common/service/scheduler/SpongeScheduler.java

```diff
@@ -0,0 +1,168 @@
+        this.asyncScheduler = new AsyncScheduler();
+        this.syncScheduler = new SyncScheduler();
+    }
+
+    public static SpongeScheduler getInstance() {
+        return INSTANCE;
+    }
+
+    @Override
+    public TaskBuilder getTaskBuilder() {
+        return new SpongeTaskBuilder();
+    }
+
+    @Override
+    public Optional<Task> getTaskById(UUID id) {
+        return this.syncScheduler.getTask(id).or(this.asyncScheduler.getTask(id));
+    }
+
+    @Override
+    public Set<Task> getTasksByName(String pattern) {
+        Pattern searchPattern = Pattern.compile(checkNotNull(pattern, "pattern"));
+        Set<Task> matchingTasks = this.getScheduledTasks();
+
+        Iterator<Task> it = matchingTasks.iterator();
+        while (it.hasNext()) {
+            Matcher matcher = searchPattern.matcher(it.next().getName());
+            if (!matcher.matches()) {
+                it.remove();
+            }
+        }
+
+        return matchingTasks;
+    }
+
+    @Override
+    public Set<Task> getScheduledTasks() {
+        Set<Task> allTasks = Sets.newHashSet();
+        allTasks.addAll(this.asyncScheduler.getScheduledTasks());
+        allTasks.addAll(this.syncScheduler.getScheduledTasks());
+        return allTasks;
+    }
+
+    @Override
+    public Set<Task> getScheduledTasks(boolean async) {
+        if (async) {
+            return this.asyncScheduler.getScheduledTasks();
+        } else {
+            return this.syncScheduler.getScheduledTasks();
+        }
+    }
+
+    @Override
+    public Set<Task> getScheduledTasks(Object plugin) {
+        String testOwnerId = checkPluginInstance(plugin).getId();
+
+        Set<Task> allTasks = this.getScheduledTasks();
+        Iterator<Task> it = allTasks.iterator();
+
+        while (it.hasNext()) {
+            String taskOwnerId = it.next().getOwner().getId();
+            if (!testOwnerId.equals(taskOwnerId)) {
+                it.remove();
+            }
+        }
+
+        return allTasks;
+    }
+
+    /**
+     * Check the object is a plugin instance.
+     *
+     * @param plugin The plugin to check
+     * @return The plugin container of the plugin instance
+     * @throws NullPointerException If the passed in plugin instance is null
+     * @throws IllegalArgumentException If the object is not a plugin instance
+     */
+    static PluginContainer checkPluginInstance(Object plugin) {
+        Optional<PluginContainer> optPlugin = Sponge.getGame().getPluginManager().fromInstance(checkNotNull(plugin, "plugin"));
+        checkArgument(optPlugin.isPresent(), "Provided object is not a plugin instance");
+        return optPlugin.get();
+    }
+
+    private SchedulerBase getDelegate(Task task) {
+        if (task.isAsynchronous()) {
+            return this.asyncScheduler;
+        } else {
+            return this.syncScheduler;
+        }
+    }
+
+    private SchedulerBase getDelegate(ScheduledTask.TaskSynchroncity syncType) {
+        switch (syncType) {
```

Comment:
This is a case of form form.

If syncType == ASYNCHRONOUS return that scheduler else return sync one. No reason to even let this be another value.